src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -494,35 +494,263 @@ private static bool HexNumberToBigInteger(ref BigNumberBuffer number, out BigInt
+                            {
+                                uint* curBufPtr = bufPtr;
+                                uint* curNewBufPtr = newBufPtr;
+                                // merge each block pairs.
+                                // When buffer represents:
+                                // |     A     |     B     |     C     |     D     |
+                                // Make newBuffer like:
+                                // |  A + B * multiplier   |  C * multiplier + D   |
+                                for (int i = 0; i < bufferSize; i += blockSize * 2)
+                                {
+                                    int len = Math.Min(bufferSize - i, blockSize * 2);
+                                    int lowerLen = Math.Min(len, blockSize);
+                                    int upperLen = len - lowerLen;
+                                    if (upperLen != 0)
+                                    {
+                                        BigIntegerCalculator.Multiply(mulPtr, blockSize, curBufPtr + blockSize, upperLen, curNewBufPtr, len);
+                                    }
+
+                                    long carry = 0;
+                                    int j = 0;
+                                    for (; j < lowerLen; j++)
+                                    {
+                                        long digit = (curBufPtr[j] + carry) + curNewBufPtr[j];
+                                        curNewBufPtr[j] = unchecked((uint)digit);
+                                        carry = digit >> 32;
+                                    }
+                                    if (carry != 0)
+                                    {
+                                        while (true)
+                                        {
+                                            curNewBufPtr[j]++;
+                                            if (curNewBufPtr[j] != 0)
+                                            {
+                                                break;
+                                            }
+                                            j++;
+                                        }
+                                    }
+
+                                    curBufPtr += blockSize * 2;
+                                    curNewBufPtr += blockSize * 2;
+                                }
+                            }
+
+                            var tmp = buffer;
+                            buffer = newBuffer;
+                            newBuffer = tmp;
+                            blockSize *= 2;
+
+                            if (bufferSize <= blockSize)
+                            {
+                                break;
+                            }
+                            newBuffer.Clear();
+                            int[]? arrayToReturn = arrayFromPool;
+
+                            arrayFromPool = ArrayPool<int>.Shared.Rent(blockSize);
```

Comment:
I don't have a sense of how many times this process runs... there's a lot of renting and returning going on.

Assuming that BigIntegerCalculator can't square in place, can we get it down to two rents and two returns by just renting two things that are big enough at the beginning, returning them at the end, and toggling between them in the square step?